### ISSUE TITLE
Some tests in k8s.io/kubernetes/pkg/kubelet/deviceplugin pkg are flaky

### ISSUE THREAD

**Issue by jiayingz** [kubernetes/kubernetes#52560]
/kind bug

**What happened**:
Some unit tests in deviceplugin/manager.go didn't clean up their states, which could cause subsequent test failures.
There is also a race in deviceplugin manager Stop that it should protect its operations with mutex.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:
go test -count 50 -race k8s.io/kubernetes/pkg/kubelet/deviceplugin

**Anything else we need to know?**:
Many thanks to @RenaudWasTaken for noticing some of these issues!

**Environment**:
- Kubernetes version (use `kubectl version`): 1.8

**Comment by mindprince** [comment]
/reopen

Lots of flakes in `TestDevicePluginReRegistration` still: https://storage.googleapis.com/k8s-gubernator/triage/index.html?test=TestDevicePluginReRegistration

According to http://velodrome.k8s.io/dashboard/db/bigquery-metrics?panelId=7&fullscreen&orgId=1 it's the flakiest unit test in the last week.

**Comment by jiayingz** [comment]
I am looking into it now.

**Comment by liggitt** [comment]
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/52950/pull-kubernetes-unit/55117/

```
=== RUN   TestDevicePluginReRegistration
2017/09/25 14:19:53 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /tmp/mock.sock: connect: no such file or directory"; Reconnecting to {/tmp/mock.sock <nil>}
2017/09/25 14:19:54 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /tmp/mock.sock: connect: no such file or directory"; Reconnecting to {/tmp/mock.sock <nil>}
2017/09/25 14:19:55 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /tmp/mock.sock: connect: no such file or directory"; Reconnecting to {/tmp/mock.sock <nil>}
2017/09/25 14:19:59 Starting to serve on /tmp/device_plugin/device-plugin.sock
2017/09/25 14:19:59 ListAndWatch
--- FAIL: TestDevicePluginReRegistration (10.02s)
	<autogenerated>:12: 
                          
	Error Trace:	manager_test.go:74
		
	Error:      	Not equal: 
		
	            	expected: 2
		
	            	actual: 0
		
	Messages:   	Devices are not updated.
```

**Comment by RenaudWasTaken** [comment]
See: https://github.com/kubernetes/kubernetes/pull/52616

**Comment by jiayingz** [comment]
This issue should be fixed with PR #53028

**Comment by RenaudWasTaken** [comment]
Please also #52616 that adds synchronous tests rather than wait based
